### PR TITLE
feat: support authentication

### PR DESCRIPTION
**Pull request by MunifTanjim**

### CODE REVIEW

**Comment by CharlesChiuGit** [comment]
wow!

File: lua/copilot/extensions/auth.lua
Comment (zbirenbaum) [inline lua/copilot/extensions/auth.lua:1]: Could you move auth to the parent dir? Extensions is primarily just extended copilot functionality

File: lua/copilot/init.lua
Comment (zbirenbaum) [inline lua/copilot/init.lua:43]: Can you move M.auth to utils instead of init.lua?

File: lua/copilot/extensions/auth.lua
Comment (zbirenbaum) [inline lua/copilot/extensions/auth.lua:1]: Could you move auth to the parent dir? Extensions is primarily just extended copilot functionality

File: lua/copilot/init.lua
Comment (zbirenbaum) [inline lua/copilot/init.lua:2]: Remove this line after moving M.auth to utils

File: lua/copilot/init.lua
Comment (zbirenbaum) [inline lua/copilot/init.lua:43]: Once you put M.auth in utils, add this to create_cmds

```
  vim.api.nvim_create_user_command("CopilotAuth", function()
    require("copilot.util").auth()
  end, {})
```

File: README.md
Comment (zbirenbaum) [inline README.md:17]: Change `:lua require("copilot").auth()` to `:CopilotAuth` after making the other changes

**Review by zbirenbaum — CHANGES_REQUESTED** [review]
Absolutely awesome PR, thank you so much for submitting this. I left a few review notes on some super minor changes just to make this integrate a bit more cleanly into the current project structure. 

To be honest, you have already done plenty here, so I am happy to merge this as is and just go in and make the changes myself right after if you would prefer that, feel free to let me know your preference after giving my comments a quick look.

File: lua/copilot/extensions/auth.lua
Comment (MunifTanjim) [reply lua/copilot/extensions/auth.lua:1]: You mean move the whole `lua/copilot/extensions/auth.lua` to `lua/copilot/auth.lua`?

File: lua/copilot/init.lua
Comment (MunifTanjim) [reply lua/copilot/init.lua:43]: Might have a chicken-egg problem adding this to `create_cmds`, (assuming `:CopilotAuth` is supposed to be callable even before setup is done... thinking of a situation where setup is done lazily when user first enters insert mode or something).

Any reason for not adding these commands to `plugin/copilot.lua`?

File: lua/copilot/extensions/auth.lua
Comment (zbirenbaum) [reply lua/copilot/extensions/auth.lua:1]: > You mean move the whole `lua/copilot/extensions/auth.lua` to `lua/copilot/auth.lua`?

Yeah. I think there should be a utils dir at some point, but for right now that is most consistent with how the project is structured and will make it easier to migrate to a directory dedicated to that later.

I know I had it in extensions, but that was mostly because it was an in progress thing and I planned to move it out once completed.

File: lua/copilot/init.lua
Comment (zbirenbaum) [reply lua/copilot/init.lua:43]: > Might have a chicken-egg problem adding this to `create_cmds`, (assuming `:CopilotAuth` is supposed to be callable even before setup is done... thinking of a situation where setup is done lazily when user first enters insert mode or something).

That's a good point. You could wrap it in its own function and call said function once things are ready.

> Any reason for not adding these commands to `plugin/copilot.lua`?
That's a good idea as long as it doesn't cause the chicken and egg problem you described. I was concerned about needing control over when the cmds were registered.

Comment (MunifTanjim) [reply lua/copilot/init.lua]: Nevermind, kept it simple for now. `:CopilotAuth` will be used by the user only once. So I don't think we need to support lazy loading right now.

If somebody lazy loaded, they'll just have to run `:CopilotAuth` once it's defined after `require("copilot").setup(...)`

File: README.md
Comment (MunifTanjim) [reply README.md]: Done.

**Comment by zbirenbaum** [comment]
@MunifTanjim Something seems to have broken, the auth window pops up, and I can enter the code on the page, but it doesn't create a folder in my config dir with the hosts.json

**Comment by MunifTanjim** [comment]
Once I see this in browser

image

It takes about 5~10 seconds to finish the whole process, and at the end of it the popup will close automatically.

I just tried again, and it seems to be working.

**Comment by zbirenbaum** [comment]
> Once I see this in browser
> 
> image
> 
> It takes about 5~10 seconds to finish the whole process, and at the end of it the popup will close automatically.
> 
> I just tried again, and it seems to be working.


Ahh, it just took a little longer that time, its working fine now. Can you add a line in the popup stating it will autoclose once completed? I'll merge it right after

**Comment by zbirenbaum** [comment]
Looks perfect! Thanks for writing this!

**Review by zbirenbaum — APPROVED** [review]

**Pull request merged**
